codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -572,46 +573,77 @@ private class ServerHttpProtocolImplGenerator(
         if (pathBindings.isEmpty()) {
             return
         }
-        val pattern = StringBuilder()
         val httpTrait = httpBindingResolver.httpTrait(operationShape)
-        httpTrait.uri.segments.forEach {
-            pattern.append("/")
-            if (it.isLabel) {
-                pattern.append("(?P<${it.content}>")
-                if (it.isGreedyLabel) {
-                    pattern.append(".+")
-                } else {
-                    pattern.append("[^/]+")
+        with(writer) {
+            val greedyLabelIndex = httpTrait.uri.segments.indexOfFirst { it.isGreedyLabel }
+            val segments =
+                if (greedyLabelIndex >= 0)
+                    httpTrait.uri.segments.slice(0 until (greedyLabelIndex + 1))
+                else
+                    httpTrait.uri.segments
+            val restAfterGreedyLabel =
+                if (greedyLabelIndex >= 0)
+                    httpTrait.uri.segments.slice((greedyLabelIndex + 1) until httpTrait.uri.segments.size).joinToString(prefix = "/", separator = "/")
+                else
+                    ""
+            val labeledNames = segments
+                .mapIndexed { index, segment ->
+                    if (segment.isLabel) { "m$index" } else { "_" }
                 }
-                pattern.append(")")
-            } else {
-                pattern.append(it.content)
+                .joinToString(separator = ",")
+            val nomParser = segments
+                .map { segment ->
+                    if (segment.isGreedyLabel) {
+                        "#{Nom}::combinator::rest"
+                    } else if (segment.isLabel) {
+                        """#{Nom}::branch::alt((#{Nom}::bytes::complete::take_until1("/"), #{Nom}::combinator::rest))"""
+                    } else {
+                        """#{Nom}::bytes::complete::tag("${segment.content}")"""
+                    }
+                }
+                .joinToString(
+                    prefix = "#{Nom}::sequence::tuple::<_, _, #{Nom}::error::Error<&str>, _>((",
+                    postfix = "))",
+                    transform = { parser ->
+                        """
+                        #{Nom}::sequence::preceded(#{Nom}::bytes::complete::tag("/"),  $parser)
+                        """.trimIndent()
+                    }
+                )
+            rustTemplate("let input_string = request.uri().path();")
+            if (restAfterGreedyLabel.isNotEmpty()) {
+                rustTemplate(
+                    """
+                    if ! input_string.ends_with(${restAfterGreedyLabel.dq()}) {
+                        return std::result::Result::Err(#{SmithyRejection}::MissingQueryString(aws_smithy_http_server::rejection::MissingQueryString));
+                    }
+                    let input_string = input_string[..(input_string.len() - ${restAfterGreedyLabel.dq()}.len())].into();
```

Comment:
It's a str without a size known at compile time https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dfafab556132a719eb28b50ab42cf70e

Reply:
`input_string` is a `&str`, `input_string[..7]` is a `str`, and the question was, why the `input_string[..7].into()`? What is it converting into? By not annotating the type of the new binding for `input_string`, it's relying on the compiler's type inference to pick an `Into` impl that makes the program compile. And it's hard to know what it's inferring from reading the code, since the next time `input_string` is used is as the input to the result of `nom::sequence::tuple`, which is very generic. Surely it must be a `&str` though, but then the call to `into()` is a bit misleading since the conversion from `str` to `&str` is very straightforward and best conveyed by simply creating a reference with `&`:

```
let input_string = &input_string[..(input_string.len() - ${restAfterGreedyLabel.dq()}.len())];
```

Reply:
I see. Yes, `&input_string[..]` also works. Updated, thanks!